### DISCUSSION TITLE
Add arabic language to doctr

### DISCUSSION THREAD

**Discussion by sridi-Rania** [mindee/doctr#1893] [General]
Hello , i'am working in my final year project and i want to fine tune doctr and add the arabic luangage to it but it doesen't work predection are always empty and i think the problem is due to the ctc decoder or I don't know can any one help me to add the arabic to it ?

**Comment by felixdittrich92** [comment]
Hi @sridi-Rania 👋🏼,
Could you provide a few more information what you have tried ? :)

**Reply by sridi-Rania** [reply]
I'm trying to add Arabic language support to **Doctr OCR**, but every time I run the model, the predicted text is empty—nothing is detected. Even when I test with images that are in the **train** folder, I still get no predictions. I'm not sure if there's a specific process required to add a new language to Doctr beyond what is mentioned in the documentation . Any guidance on this would be greatly appreciated  even if you can tell me about the steps that i should follow to have a result ! i can show you an exemple when i'am trying to train the model 
![image](https://github.com/user-attachments/assets/d7b3ce90-26e2-41d0-bc7d-2500d2147fa6)

**Reply by felixdittrich92** [reply]
So to break this done:

First: Your model doesn't learns anything the loss is much to high and swaps randomly

Do you use the internal provided vocab `VOCABS['arabic']` ? (Doesn't looks so because our provided one has a size of 101 but you provide a vocab with 426 chars + EOS / PAD / SOS

How much samples are in your training / val sets ?

Which arguments do you use for training ?


In general all our models are pretrained on the french (latin ext) vocab which works pretty well for fine tuning on other latin based vocabs - But for other writing systems like chinese / arabic / hindi and so on I think the only way would be to train from scratch which requires lot's of data (could be 100% synth for pretraining) - our models was pretrained on ~10M real word crops

We have already planned to make docTR multilingual but this will require some time on our end ref.: #1699 

CC @sebastianMindee

**Reply by sridi-Rania** [reply]
I have changed the vocabulary because every time I run the code, an error occurs, stating that some characters are not in the Arabic vocabulary.  

I just want to test whether the fine-tuning works or not, so I used a very small dataset containing 1000 examples (800 samples for training and 200 for validation). The issue might be caused by the small dataset. ? sorry for bothering

**Reply by felixdittrich92** [reply]
That's to less yes ..but additional which chars are missing in our predefined arabic vocab ? :)

Your labels shouldn't contain whitespaces and 1 word/image and the depending string (for example: "Hello")

**Reply by felixdittrich92** [reply]
Reference datasets can be found here: https://github.com/mindee/doctr/discussions/1654

**Reply by sridi-Rania** [reply]
In fact, Arabic letters have multiple forms depending on their position in a word. However, your Arabic vocabulary only includes the basic forms of the letters, making it incomplete. i can take an exemple these are ﺀ ﺁ ﺃ ﺅ ﺇ ﺉ  all  forms for the same alphabet

**Reply by felixdittrich92** [reply]
At the end it depends how utf-8 encodes the letters for example for `gujarati` we have had to use some "variables"  because some letters was a combination of the basic letter and the circle dotted "extension" -> so one letter on paper but read by the system two.

```
"gujarati_vowels": "અઆઇઈઉઊઋએઐઓ",
"gujarati_consonants": "ખગઘચછજઝઞટઠડઢણતથદધનપફબભમયરલવશસહળક્ષ",
"gujarati_digits": "૦૧૨૩૪૫૬૭૮૯",
"gujarati_punctuation": "૰ઽ◌ંઃ॥ૐ઼ઁ" + "૱",
```

The easiest would be to read all your labels in a list, extract all unique chars and deduplicate then we could update our internal vocab wdyt ? :)

**Reply by sridi-Rania** [reply]
Yes, I've done that and increased the dataset size to 20,000, but I think the model still isn't learning properly. The issue might be due to the greedy CTC decoder used in the model ?
![Capture d'écran 2025-03-13 113123](https://github.com/user-attachments/assets/8a61cf87-87c7-4873-bb75-357a2a487739)

**Reply by felixdittrich92** [reply]
But loss looks much better now :+1: 

How does it look if you add `--pretrained` to your training command (for fine tuning) ? Because actual you train from scratch where 20k is of course to less data

**Reply by sridi-Rania** [reply]
I'll give it a try and get back to you with either success or failure. Thank you!

**Comment by sridi-Rania** [comment]
Hello Is it possible to freeze all layers except one for fine-tuning, since my dataset is small?

**Reply by felixdittrich92** [reply]
Hi @sridi-Rania :wave: 

This would require that you modify the training script slightly:

```
children = list(model.children())

for child in children[:-1]:
    for param in child.parameters():
        param.requires_grad = False
```

This will freeze everything except the last linear layer (Attention: It depends on the model you want to fine-tune because `master` `sar_resnet31` and `parseq` uses an Embedding layer for these models you need to keep also the embedding layer trainable)

Untested but the following should work for all models:

```
# Freeze everything first
for param in model.parameters():
    param.requires_grad = False

# Unfreeze embedding layer
for module in model.modules():
    if isinstance(module, nn.Embedding):
        for param in module.parameters():
            param.requires_grad = True

# Unfreeze last linear (head)
linear_layers = [m for m in model.modules() if isinstance(m, nn.Linear)]
if linear_layers:
    for param in linear_layers[-1].parameters():
        param.requires_grad = True
```

Best,
Felix

**Reply by felixdittrich92** [reply]
For backbone (feature extractor) freezing we have an option `--freeze-backbone`

**Reply by sridi-Rania** [reply]
Hello, I’d like to ask something. I’m currently training the recognition model parseq using a dataset where each image contains a single word. Should I continue with this approach, or would it be better to train the model using images that contain full phrases? thank you for your help

**Reply by felixdittrich92** [reply]
Only 1 word per image we don't support lines or phrases so everything correct :+1: 
Btw. you mentioned that our arabic vocab isn't complete what's missing ? :)

**Reply by sridi-Rania** [reply]
Yes, in fact, in our Arabic language, each letter can take on multiple forms depending on its position within a word. I can provide you with the complete set of letter forms if needed. I'm not sure if it's the correct approach, but when I tried to train the model using the vocabulary you provided, it didn’t work. So I had to include the Arabic letters in the vocabulary file separated by spaces, in order to represent all their contextual forms and this is the only way that make the train work

**Reply by felixdittrich92** [reply]
Yeah this would be great :+1: 

At the end we should cover each unique "char" in our vocab so that we later on can generate each combination vowel + diacritic combination for example

**Reply by sridi-Rania** [reply]
This file contains all contextual forms of the Arabic alphabet depending on their position in words. You can check it — it might include some unnecessary forms or something like that. 
[alph.txt](https://github.com/user-attachments/files/20081245/alph.txt)

**Comment by sridi-Rania** [comment]
I trained the model, but despite the values and precision, it failed to detect anything when I used photos from the internet. I honestly don't know what to do. Could the issue be related to the fonts? Should I try using different fonts instead of the default ones? Do you have any suggestions
![image](https://github.com/user-attachments/assets/04241d88-28ae-4e34-afd2-4d799df287b3)

**Reply by felixdittrich92** [reply]
Hey :)
Which tool have you used to generate the dataset ?

**Reply by felixdittrich92** [reply]
I collected already some fonts can be found here https://huggingface.co/datasets/Felix92/docTR-resource-collection/tree/main/sorted_fonts/arabic

**Reply by sridi-Rania** [reply]
I generated the data using code that I wrote myself, and I also collected some Arabic fonts.

**Reply by felixdittrich92** [reply]
Could you share such a synth generated sample ?
I use atm a modified version of SynthTiger which works pretty well (only tested with latin extended / cyrillic extended / greek / hebrew)

Do you have a prepared wordlist which covers the whole vocab you want to recognize ?

**Reply by sridi-Rania** [reply]
This is an example of the generated data:
![image](https://github.com/user-attachments/assets/2e849d36-6e87-45ee-bd36-46b5560b5304)
I prepared a .txt file containing Arabic words, and I used it to generate the images I can show you an exemple 
[texte1.txt](https://github.com/user-attachments/files/20141370/texte1.txt)

**Reply by felixdittrich92** [reply]
The image should contain only the text without the huge padding keep in mind for recognition every image will be resized to 32x128 (HxW) :)

**Reply by sridi-Rania** [reply]
So the problem could be caused by this issue? But can I change the size? Because 32×128 can't fully contain the word — it gets cut off in the middle

**Reply by felixdittrich92** [reply]
That's an issue yes but in general I would suggest to generate a more various dataset that the model can learn to generalize

(Not tested for arabic yet) I used the following (dirty) modified SynthTiger version https://github.com/felixdittrich92/synthtiger/tree/doctr-modified (branch: doctr-modified)

Would be also curious to see if that works so if you could share your complete wordlist I could generate a dataset (next week) from it and share. Let me know :)

**Reply by felixdittrich92** [reply]
> So the problem could be caused by this issue? But can I change the size? Because 32×128 can't fully contain the word — it gets cut off in the middle

Mh you could change the input size but I wouldn't suggest to do this .. It doesn't needs to be exactly 32x128 but something around (64x200, 24x180, ...)

**Reply by sridi-Rania** [reply]
I generated this image with a size of 128×32. What do you think? Should I change any parameters or something else?
![image](https://github.com/user-attachments/assets/99f49a9a-170d-4475-85ea-fbff9befc911)

**Reply by felixdittrich92** [reply]
Looks better but still to much white border (right & left side) it should really only contain the word

**Reply by sridi-Rania** [reply]
Hello again,
I’ve managed to get a decent result for the recognition part. However, I’m now facing issues with detection. Ideally, each page should be processed with bounding boxes around each word, but the detection model isn’t performing well, even after fine-tuning.
Is it possible to fine-tune the model using images that contain multiple words, with each word annotated with its own bounding box? Or do you have any suggestions or best practices to improve word-level detection?
Thank you in advance for your help!

**Reply by felixdittrich92** [reply]
Hey great to hear that it works better now :)

We have an small OS tool you could use to annotate some samples for fine-tuning: [doctr-labeler](https://github.com/text2knowledge/docTR-Labeler).

With just ~200 samples (80/20 split), we already achieved a noticeable improvement.

This will at the end export the data directly in the doctr required format for fine-tuning:
https://github.com/mindee/doctr/tree/main/references/detection#readme

**Reply by sridi-Rania** [reply]
Thank you , I will update you if it works

**Reply by felixdittrich92** [reply]
4-point polygons (boxes) should be close as possible to the text - by default after drawing a box it will try to auto-correct the box this needs sometimes a bit manual adjustment afterwards :)

**Reply by felixdittrich92** [reply]
If there are any issues or questions please open it in the doctr-labeler repo :+1:

**Reply by sridi-Rania** [reply]
I have a question about the format of the data for the detection model. When I have more than one word in the image and I save the polygons in the JSON file, I see in the structure 'class name 1', 'class name 2', etc. What is meant by 'class name 1', 'class name 2', etc.
I'm really sorry for the disturbance. 
![image](https://github.com/user-attachments/assets/e187696b-78bc-4b8a-8400-b70a8df6f60f)

**Reply by felixdittrich92** [reply]
Ah no that's for kie.. you need the following format:

```json
{
    "sample_img_01.png" = {
        'img_dimensions': (900, 600),
        'img_hash': "theimagedumpmyhash",
        'polygons': [[[x1, y1], [x2, y2], [x3, y3], [x4, y4]], ...]
     },
     "sample_img_02.png" = {
        'img_dimensions': (900, 600),
        'img_hash': "thisisahash",
        'polygons': [[[x1, y1], [x2, y2], [x3, y3], [x4, y4]], ...]
     }
     ...
}
```

**Reply by sridi-Rania** [reply]
Hello again, while trying to fine-tune the detection model with 200 examples, it didn't work. Should the images have a specific size or format, or could there be another issue? , thank you

**Reply by felixdittrich92** [reply]
Hi :)

Could you share the training command you used and if possible an example with displaying the annotations ?

under the hood everything is resized to 1024x1024 by keeping the aspect ratio and symmetric padding

**Reply by felixdittrich92** [reply]
you can start the training with `--show-samples` and share the image

**Reply by sridi-Rania** [reply]
HEllo again , sorry  i resolved the problem but i have just a question when training the recognition model with the arabic data , I used the 'multilingual' option as an argument and a pretrained model. However, after training, when testing the model, it does not recognize the other languages I had specified. Could you please correct this issue?

**Reply by felixdittrich92** [reply]
Hey :)
It can only recognize the chars which are in your training dataset everything else will be "forgetten" because the head layer is completly reset

**Comment by felixdittrich92** [comment]
Reference datasets can be found here for both recognition and detection :)
https://github.com/mindee/doctr/discussions/1654

**Comment by sridi-Rania** [comment]
hope this message finds you well. I am pleased to share that I have successfully fine-tuned the Doctr model for the Arabic language, and it is now performing well in this context.
I would be honored to contribute my model to your GitHub repository and become an active contributor to the project. I am also willing to support other users who are interested in working with Arabic, and I look forward to helping expand Doctr's multilingual capabilities.
Thank you for considering my request. I look forward to your response.

**Reply by felixdittrich92** [reply]
Hi @sridi-Rania :wave:

Great to hear it’s working well now ! :+1:
We’re always happy to welcome external contributors and those sharing their models - thank you! \:hugs:

The easiest way to share your model(s) is through our Hugging Face integration:
👉 [https://mindee.github.io/doctr/using\_doctr/sharing\_models.html](https://mindee.github.io/doctr/using_doctr/sharing_models.html)

For example:

```python
from doctr.models import recognition, login_to_hub, push_to_hf_hub

login_to_hub()

my_vocab = "abc"  # use the same vocab your model was trained on
my_awesome_model = recognition.crnn_vgg16_bn(pretrained=False, pretrained_backbone=False, vocab=my_vocab)

# Load your checkpoint
reco_params = torch.load('<path_to_pt>', map_location="cpu")
my_awesome_model.load_state_dict(reco_params)

# Push to the Hub
push_to_hf_hub(
    my_awesome_model,
    model_name='doctr-crnn-vgg16-bn-arabic-v1',
    task='recognition',
    arch='crnn_vgg16_bn'
)
```

If you’re interested in contributing to docTR, feel free to join our [Slack community](https://mindee-community.slack.com/signup#/domain-signup) for easier communication - I’ll ping you once you’ve joined! 😊

Best,
Felix

**Reply by sridi-Rania** [reply]
I would like to contribute to the Doctr project and join the Mindee Community workspace. 
Could you please send me an invitation to join the workspace? You can use this email address for the invitation: raniasridi25@gmail.com, as I am unable to join on my own — the platform requires a @mindee.com email address.
Thank you very much for your time and consideration

**Reply by felixdittrich92** [reply]
Done :)
